File: 134-gas-station/gas-station.py
class Solution(object):
    def canCompleteCircuit(self, gas, cost):
        """
        :type gas: List[int]
        :type cost: List[int]
        :rtype: int
        """
        n = len(gas)

        currentGas = 0
        cst = 0
        netGas = [0] * n
        for i in range(n):
            currentGas += gas[i]
            cst += cost[i]
            netGas[i] = gas[i] - cost[i]

        if cst > currentGas:
            return -1
        
        ans = -1
        total = 0
        for i in range(n):
            if total + netGas[i] < 0:
                ans = -1
                total = 0
            else:
                total += netGas[i]
                if ans == -1:
                    ans = i
    
        return ans

File: 134-gas-station/gas-station.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int n = gas.size();

        int currentGas = 0, cst = 0;
        vector<int> prefixSum(n, 0);
        for(int i = 0; i < n; i++){currentGas += gas[i]; cst += cost[i]; prefixSum[i] = gas[i] - cost[i];}

        if(cst > currentGas)
            return -1;
        int ans = -1;
        int total = 0;
        for(int i =0 ; i < n; i++){
            if(total + prefixSum[i] < 0){
                ans = -1; 
                total = 0;
            }
            else{
                total += prefixSum[i];
                if(ans == -1)
                    ans = i;
            }
        }
        
        return ans;
    }
};